dist/clouduploader.rb

```diff
@@ -4,20 +4,86 @@
 require 'json'
 require 'ostruct'
 require 'open3'
+require 'fileutils'
 
 module CloudUploader
+  # Module method for uploading the image depending on the platform
+  def self.upload(user, platform, backend_image_file, job_data_file, image_filename, result_path)
+    STDOUT.sync = true
+    start = Time.now
+
+    job_data = JSON.parse(File.read(job_data_file))
+
+    case platform
+    when 'ec2'
+      FileUtils.ln_s(backend_image_file, File.join(Dir.pwd, image_filename)) # link file into working directory
+      EC2.upload(image_filename, job_data, result_path)
+    when 'azure'
+      Azure.upload(backend_image_file, image_filename, job_data, result_path)
+    else
+      abort('No valid platform. Valid platforms are ec2 and azure.')
+    end
+
+    diff = Time.now - start
+    STDOUT.write("Upload took: #{Time.at(diff).utc.strftime("%H:%M:%S")}\n")
+  end
+
   class EC2
+    def self.upload(image, data, result_path)
+      ENV['PYTHONUNBUFFERED'] = '1'
+
+      STDOUT.write("Start uploading image #{image}.\n")
+
+      credentials = credentials(data)
+
+      command = [
+        'ec2uploadimg',
+        "--description='obs uploader'",
+        '--machine=x86_64',
+        "--name=#{data['ami_name']}",
+        "--region=#{data['region']}",
+        "--secret-key=#{credentials.secret_access_key}",
+        "--access-id=#{credentials.access_key_id}",
+        "--session-token=#{credentials.session_token}",
+        '--verbose'
+      ]
+
+      if data['vpc_subnet_id']
+        command << "--vpc-subnet-id=#{data['vpc_subnet_id']}"
+      end
+      command << image
+
+      Open3.popen2e(*command) do |_stdin, stdout_stderr, wait_thr|
+        Signal.trap("TERM") {
+          # We just omit the SIGTERM because otherwise we would not get logs from ec2uploadimg
+          STDOUT.write("Received abort signal, waiting for ec2uploadimg to properly clean up.\n")
+        }
+        while line = stdout_stderr.gets
+          STDOUT.write(line)
+          write_result($1, result_path) if line =~ /^Created\simage:\s+(ami-[\w]+)$/
+        end
+        status = wait_thr.value
+        abort unless status.success?
+      end
+    end
+
+    def write_result(result, result_path)
+      File.open(result_path, "w+") { |file| file.write(result) }
+    end
+
     def self.credentials(data)
-      # Credentials are stored in  ~/.aws/credentials
-      out, err, status = Open3.capture3(
+      command = [
         'aws',
         'sts',
         'assume-role',
         "--role-arn=#{data['arn']}",
         "--external-id=#{data['external_id']}",
         '--role-session-name=obs',
         "--duration-seconds=#{THIRTY_MINUTES}"
-      )
+      ]
+
+      # Credentials are stored in  ~/.aws/credentials
+      out, err, status = Open3.capture3(*command)
```

Comment:
Did you try that? (I think I tried it and did not work)

Reply:
I don't get what you mean...